File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-13/Estruturais/Composite/Composite01.py
# Classe Componente:
class Componente:
    def __init__(self, nome):
        self.nome = nome

    def adicionar(self):
        pass

    def remover(self):
        pass

    def obterNome(self):
        pass

class Folha(Componente):
    def __init__(self, nome, preco):
        super().__init__(nome)
        self.preco = preco
    
    def obterNome(self):
        return self.nome

    def obterPreco(self):
        return self.preco
    
class Container(Componente):
    def __init__(self, nome):
        super().__init__(nome)
        self.componentes = []

    def adicionar(self, componente):
        self.componentes.append(componente)

    def remover(self, componente):
        index = self.componentes.index(componente)
        self.componentes.pop(index)
    
    def obterNome(self):
        return self.nome

    def obterPreco(self):
        preco = 0
        for componente in self.componentes:
            preco += componente.obterPreco()
        return preco

# Interface de uso -------------------------------------

# Criando os elementos Frutas ----------
maca = Folha('Maçã', 4.50)
laranja = Folha('Laranja', 2.90)
uva = Folha('Uva', 5.00)
abacaxi = Folha('Abacaxi', 5.00)

# Criando os elementos Embalagem -------
embalagem = Folha('Embalagem', 4.00)

# Criando Tabela de Frete --------------
freteSP = Folha('SP', 10.00)
freteRJ = Folha('RJ', 20.00)
freteMG = Folha('MG', 30.00)

# Criando os Conteiners:
frutas = Container('Frutas')
frutas.adicionar(maca)
frutas.adicionar(laranja)
frutas.adicionar(uva)
frutas.adicionar(abacaxi)

caixa = Container('Caixa de Frutas')
caixa.adicionar(frutas)
caixa.adicionar(embalagem)

frete = Container('Frete - Entrega')
frete.adicionar(caixa)
frete.adicionar(freteEscolhido)

print(f"Preço Frutas (Apenas as Frutas)....................... R$ {frutas.obterPreco()}")
print(f"Preço Caixa de Fruta (Frutas + Embalagem)..............R$ {caixa.obterPreco()}")
print(f"Preço Frete (Caixa de Fruta + Frete) - Preço Total.....R$ {frete.obterPreco()}")

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-13/Estruturais/Composite/Composite01.js
// Classe Componente:
class Componente{
    constructor(nome){
        this.nome = nome;
    }

    adicionar(){};

    remover(){};

    obterNome(){};
}

class Folha extends Componente{

    constructor(nome, preco){
        super(nome);
        this.preco = preco;
    }

    obterNome(){
        return this.nome;
    }

    obterPreco(){
        return this.preco;
    }
}

class Conteiner extends Componente{
    constructor(nome){
        super(nome);
        this.componentes = [];
    }

    adicionar(componente){
        this.componentes.push(componente);
    }

    remover(componente){
        const index = this.componentes.indexOf(componente);
        this.componentes.splice(index, 1);
    }

    obterNome(){
        return this.nome;
    }

    obterPreco(){
        let preco = 0;
        this.componentes.forEach(componente => {
            preco += componente.obterPreco();
        });
        return preco;
    }

}

// Interface do Usuario - Utilização do Padrão:
const maca = new Folha('Maçã', 4.50);
const laranja = new Folha('Laranja', 2.90);
const uva = new Folha('Uva', 5.00);
const abacaxi = new Folha('Abacaxi', 5.00);

// A embalagem ao inves de criar como uma variavel, foi inserido direto (abaixo)
const embalagem = new Folha('Embalagem', 4.00);

// Tabela de Frete:
const freteSP = new Folha('SP', 10.00);
const freteRJ = new Folha('RJ', 20.00);
const freteMG = new Folha('MG', 30.00);

const frutas = new Conteiner('Frutas');
frutas.adicionar(maca);
frutas.adicionar(laranja);
frutas.adicionar(uva);
frutas.adicionar(abacaxi);

const caixa = new Conteiner('Caixa de Frutas');
caixa.adicionar(frutas);
caixa.adicionar(embalagem);
//caixa.adicionar(new Folha('Embalagem', 4.00));

const frete = new Conteiner('Frete');
frete.adicionar(caixa);
frete.adicionar(freteSP);

console.log(`Preço Frutas (Apenas as Frutas): ${frutas.obterPreco()}`);
console.log(`Preço Caixa de Fruta (Frutas + Embalagem): ${caixa.obterPreco()}`);
console.log(`Preço Frete (Caixa de Fruta + Frete): `)
console.log(`Preço Total: R$ ${frete.obterPreco()}`);

